lib/pbench/cli/agent/commands/tools/clear.py

```diff
@@ -0,0 +1,167 @@
+"""
+pbench-clear-tools
+
+This script will remove tools that have been registered.  If no options are
+used, then all tools from the "default" tool group are removed.  Specifying
+a tool name and/or remote host will limit the scope of the removal.
+
+
+"""
+import shutil
+import sys
+import os
+import pathlib
+
+import click
+
+from pbench.agent.utils import setup_logging
+from pbench.cli.agent import CliContext, pass_cli_context
+from pbench.cli.agent.commands.tools.base import ToolCommand
+from pbench.cli.agent.options import common_options
+
+
+class ClearTools(ToolCommand):
+    def __init__(self, context):
+        super().__init__(context)
+
+        self.logger = setup_logging(logfile=self.pbench_log)
+
+    def execute(self):
+        """Clear registered tools"""
+        self.tools_group_dir = self.tool_group_dir(self.context.group)
+        if not self.tools_group_dir.exists():
+            return 1
+
+        try:
+            self.context.remote = self.context.remote.split(",")
+        except Exception:
+            # We were not given any remotes on the command line, build the list
+            # from the tools group directory.
+            self.context.remote = self.remote(self.tools_group_dir)
+
+        for remote in self.context.remote:
+            tg_dir = self.tools_group_dir / remote
+            if not tg_dir.exists():
+                # Ignore any specified remote that does not exist.
+                self.logger.warn(
+                    f'The given remote host, "{remote}", is not a directory in {self.tools_group_dir}.'
+                )
+                continue
+
+            if self.context.name:
+                # Use the given tool name from the command line
+                self.context.name = [self.context.name]
+            else:
+                # Discovery all the tools registered for this remote
+                self.context.name = self.tools(tg_dir)
+
+            for name in self.context.name:
+                self._clear_tools(name, remote.name)
+
+            tool_files = [p.name for p in tg_dir.iterdir()]
+
+            if "__label__" in tool_files:
+                label = tg_dir / "__label__"
+                try:
+                    os.unlink(label)
+                except OSError:
+                    self.logger.error(f"Failed to remove label for remote {tg_dir}")
+                    return 1
```

Comment:
The `bash` code doesn't quit after logging the error; it seems to count the error and continue... but it also wipes out `tools_files` which would allow the subsequent `rmdir` even if the `unlink` failed. This version won't process the rest of the remotes if that happens...

Reply:
Fixed